### PR TITLE
GetIdFromArguments now tries to cache consistent results - saves 0.3s of init time

### PR DESCRIPTION
**Pull request by Mothblocks**
Elements now attempt to cache their ID per type, under the assumption that most elements are added with consistent arguments.

### CODE REVIEW

**Comment by Mothblocks** [comment]
Ninja gave a suggestion like this but I'm not sure if the memory costs of this cancel out elements in general?

**Comment by Mothblocks** [comment]
Oh that is bad, it's hard-deling with exact references...I'm surprised any are being held.

Will probably need to scan for any `isdatum`.

**Comment by Mothblocks** [comment]
Oh wait no it's because I'm not snipping the arguments

File: code/controllers/subsystem/dcs.dm [hunk 1]
```diff
@@ -57,4 +85,16 @@ PROCESSING_SUBSYSTEM_DEF(dcs)
...
 		fullid += named_arguments
 
-	return list2params(fullid)
+	var/id = list2params(fullid)
+
+	if (isnull(cached_id_info))
+		if (isnull(ids_per_type[eletype]))
+			ids_per_type[eletype] = list((source_path) = list(arguments.Copy(), id))
+		else
+			ids_per_type[eletype][source_path] = list(arguments.Copy(), id)
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/dcs.dm:94]: what happens if arguments contains a datum. will it hang a ref?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/dcs.dm (hunk 1 above)
Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:94]: Yes, I think we're going to find out with create and destroy if anything does that. Though I can't think of a valid element that would do that?

File: code/controllers/subsystem/dcs.dm [hunk 2]
```diff
@@ -8,29 +8,57 @@ PROCESSING_SUBSYSTEM_DEF(dcs)
...
+
+	var/list/cached_id_info = ids_per_type[eletype]?[source_path]
+	if (islist(cached_id_info))
+		var/list/segments = cached_id_info[CACHED_ID_INFO_KEY_SEGMENTS]
+		var/segments_length = length(segments)
+
+		if (segments_length == length(arguments))
+			var/all_valid = TRUE
+
+			for (var/index in initial(eletype.id_arg_index) to segments_length)
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:53]: 241k calls in a second, it's one instruction

File: code/controllers/subsystem/dcs.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/controllers/subsystem/dcs.dm:94]: a bespoke element could pass in a second object to register to, and increment the cache index.
That's what comes directly to mind. feels silly to break that behavior

File: code/controllers/subsystem/dcs.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/controllers/subsystem/dcs.dm:53]: I asked cause you could do this by going back to front, which if it was expensive would save some time

File: code/controllers/subsystem/dcs.dm (hunk 1 above)
Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:94]: I'll do a scan through of all new elements before putting into the cache

File: code/controllers/subsystem/dcs.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:53]: Let's kill it and make it a constant 2 😃

File: code/controllers/subsystem/dcs.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/controllers/subsystem/dcs.dm:94]: would it be possible to only save the arguments that are bespoke?
then we can just count backwards through both lists
we just don't get the best case optimization as a side effect, and potentially hang refs to bespoke args

Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:94]: I don't know what you mean by that sorry

Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:94]: @LemonInTheDark What do you mean by that

Comment (LemonInTheDark) [reply code/controllers/subsystem/dcs.dm:94]: I don't like that we can currently save non bespoke args in our cache list
I would prefer to instead just the bespoke arguments, and then walk both lists backwards to compare them.
this means the length comparison either requires a cut, or we need to discard it and walk the full list each time.

I could be wrong about that tho, think I am. but that's what I was thinking when I said it

Comment (Mothblocks) [reply code/controllers/subsystem/dcs.dm:94]: I don't see how walking backwards is going to help anything if the lengths are equivalent

**Comment by Mothblocks** [comment]
@LemonInTheDark I can't finish this right now but you should look into this

**Comment by LemonInTheDark** [comment]
barrel

**Pull request closed without merging**